src/Metadata/Driver/AttributeDriver/AttributeReader.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+namespace JMS\Serializer\Metadata\Driver\AttributeDriver;
+
+use Doctrine\Common\Annotations\Reader;
+use ReflectionClass;
+use ReflectionMethod;
+use ReflectionProperty;
+
+class AttributeReader implements Reader
```

Comment:
I feel the 3 pairs of methods (for class, for method and for property) are doing all the same, don't they?

If this is due to copy/paste and not a will to let the possibility of each way evolve in their own ways, this could be easily reduced by 2 private (protected?) methods.

```php
    private function buildAnnotation($attributes)
    {
        if (!isset($attributes[0])) {
            return null;
        }
        return $attributes[0]->newInstance();
    }

    private function buildAnnotations($attributes)
    {
        $result = [];
        foreach ($attributes as $attribute) {
            if (0 === strpos($attribute->getName(), 'JMS\Serializer\Annotation\\')) {
                $result[] = $attribute->newInstance();
            }
        }
        
        return $result;
    }
```

And then:
```php
    public function getClassAnnotations(ReflectionClass $class)
    {
        $attributes = $class->getAttributes();
        return $this->buildAnnotations($attributes);
    }
    
    public function getClassAnnotation(ReflectionClass $class, $annotationName)
    {
        $attributes = $class->getAttributes($annotationName);
        return $this->buildAnnotation($attributes);
    }
    
    // Apply the same refactor for getMethodAnnotation(s) and getPropertyAnnotation(s)
```

Otherwise, if you want to keep the idea of making every method to evolve in their own ways independently, then you can keep it as is for now, but you should have 6 classes to split the responsibilities instead of a single one.